keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1387 @@
+management operations by a container runtime are:
+- New field in `RuntimeClass` to specify a unix domain socket path over which the
+Kubelet can invoke FileSystem Management APIs that will be handled by the
+container runtime.
+- New field in CSI `NodeGetVolumeStatsRequest` and `NodeVolumeExpandRequest` for the
+Kubelet to indicate to a CSI plugin that the pod (that a PV is published to) is
+associated with a runtime that supports management of the filesystem associated
+with the PV.
+- New field in CSI `NodeGetVolumeStatsResponse` and `NodeVolumeExpandResponse` for
+a CSI plugin to indicate to the Kubelet that it should pass the request to the
+container runtime by specifying a `source` field corresponding to the volume
+(e.g. host path for a block device or server export path for a shared file system).
+- A new CSI service: `Runtime` initially with two RPCs: `RuntimeGetFileSystemStats`
+and `RuntimeExpandVolume` corresponding to CSI `NodeGetVolumeStats` and
+`NodeExpandVolume`.
+- Enhancements to Kubelet to populate the new field in CSI `NodeGetVolumeStatsRequest`
+and `NodeVolumeExpandRequest` (based on `RuntimeClass` field), extract the new
+`source` field from `NodeGetVolumeStatsResponse` and `NodeVolumeExpandResponse` and
+invoke the corresponding new `Runtime` APIs on the container runtime handler.
+- Enhancements to OCI runtimes to support CSI `Runtime` API and surface a unix
+domain socket over which the API will be invoked.
+- Enhancements to CSI plugins to inspect the new field in CSI
+`NodeGetVolumeStatsRequest` and `NodeVolumeExpandRequest` and specify deferral of
+stats and online file system expansion by populating the new `source` field
+corresponding to the volume in CSI `NodeGetVolumeStatsResponse` and
+`NodeVolumeExpandResponse`. The desired size of the file system will also need to
+be populated by the CSI plugin in `NodeVolumeExpandResponse` (if it doesn't already
+populate it).
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A pod specifying a microvm runtime (that is enabled to handle mounting and management
+of file system operations) and persistent volumes hosted in a block
+device managed by a CSI plugin (that can coordinate file system operations with
+the runtime handler) is scheduled on a node. Here are the sequence of steps that
+takes place in the context of PVs:
+- The node CSI plugin receives a CSI `NodeVolumeStage` call. The CSI plugin will
+stage the volume (ensuring it is formatted) and mount the filesystem
+(associated with the PV) in the node host OS environment. This stage happens
+irrespective of the runtime's capabailities indicated in the subsequent CSI
+NodeVolumePublish call.
```

Comment:
What's the advantage of putting this work into `NodeStageVolume`? Why not skip the call and do the work in `NodePublishVolume`?

Reply:
I will put a note clarifying ^ is the preferred approach.